### ISSUE TITLE
Javascript

### ISSUE THREAD

**Issue by gcstaublin** [gcstaublin/pregame#56]
Our initial thought was that we'd not deal with any Javascript, but there's a few places where providing universal JS might make sense.
- Float labels for forms
- RWD tables (if we go this route) http://zurb.com/playground/responsive-tables
- Error States (if we go that route)
- Dismissing banners
- Activating/Closing Modals

The reason to do this is to provide the one way that any of this interactivity occurs, versus everyone rolling their own solutions.

The question is how would this universal JS file sit within React pages. Would we need to build the components with that interactivity built in, at which point that cluster is no longer including that universal js file?

**Comment by chrisronline** [comment]
React will solve this problem well because the markup is included in the same file as the functionality. We'll probably have to work with web platform to figure out how to support backbone/marionette/handlebars components, but I think the shift to React is more official now and at some point, we can expose all components in React only and easily bundle markup and logic together.

**Comment by nick-adams** [comment]
Hmm. Some of the javascript listed there just belongs in the component itself like Chris said, which is aided by React.

The things that we didn't consider components like Forms and Tables is where this kind of universal javascript is needed. I wonder the best way to do this. Dilemma...

Maybe we have a kickoff-base.js too? Not sure, thoughts?

**Comment by chrisronline** [comment]
I'm not following - what's the specific use case with Forms or Table? I imagine we'll expose both of those as separate components with all markup/logic bundled together too.

**Comment by gcstaublin** [comment]
Forms use the `float label` pattern, when the input if focus or active the label floats above the input.

And for tables, this is a potentially nice solution, which would require JS. http://zurb.com/playground/responsive-tables

Defaulting to "React will solve this" doesn't solve these questions for the vast majority of the app. That is, until there's an edict to refactor the entire world for react. And only solving this bit of universal interactivity for only those users, seems less than ideal to me.

**Comment by chrisronline** [comment]
Ah. Yup I agree. We'll have to put this "universal javascript" (or legacy support) in [js-hudl-base](https://github.com/hudl/js-hudl-base)

**Comment by nick-adams** [comment]
That is a terrible name for that package.

**Comment by gcstaublin** [comment]
I can I put you down for the OKR to rename that, @nick-adams? :)

**Comment by nick-adams** [comment]
For real. I know this isn't on a roadmap, or an OKR, but it exposes something we'll need to address. Organizing and making sense of all these packages on the front end. Right now, it's really hard to follow the thread of all this architecture. 

We need to make a good pass on cleaning it up, adding some logical naming patterns, and defining the responsibilities of each. What contains kickoff stuff? What contains client-side MVC boilerplate stuff? What do we call each? When do we use each? etc. It all seems a bit to haphazard at the moment.

That goes all the way down to refining any kind of master/global view template for the app which is going to be a dependency of starting to implement any new navigation -- especially if they want A/B testing (they will).

A good thought exercise to test this is, why does this universal kickoff-base js going in js-hudl-base? and would people be able to know that's where it is? And why wouldn't it just be included in the kickoff repo itself? With kickoff-base being a package that loads kickoff-base.css and kickoff-base.js?

There needs to be a bit more logic to all of this.

**Comment by chrisronline** [comment]
That's a fair point - it does make more sense for us to own our javascript so disregard my previous suggestion.

`would people be able to know that's where it is` - this isn't really an issue because we have our `multiverse-template` app

**Comment by gcstaublin** [comment]
This notion didn't get much traction last time I brought it up, but in these waning days of 2015, I'll mention it one last time. 

When I hear the words Kickoff, I don't instantly know what the person is talking about. For designers the chances are when they says Kickoff, they mean the CSS/Design system where as Devs might mean something entirely different. While we've made the improvement of calling our css `kickoff-base`, conversationally folks are not going to refer to kickoff-base, they going to say kickoff.

I still wonder if the css/js thing we make is named something something different. Something that doesn't have the words `kick` or `off` in it. Hold on, stay with me for a minute.

If our CSS/JS had it's own unique name, this solves a couple problems
- It's no longer tied to a library that didn't really solve the needs of designers and developers.
- There's a clear line when someone has chosen to use our new library. The `ko-` prefix will slowly disappear from the markup, which means legacy overwrites and/or style collisions between universal and localized css will no longer exist.
- As clusters adopt this new library, the coding patterns are going to approved patterns that others can look to for their own clusters.
- This library stands on it's own, whether that mass success or moderate failure. It fails under it's own name and not under what I would, without hyperbole, say is a tainted name.

That's my final pitch*

* for 2015.

**Comment by chrisronline** [comment]
I'm not against changing the name, but it _has_ to be better. What names do you like? :)
